File: scripts/fix-console-logs.py
#!/usr/bin/env python3
"""
Replace console.log with structured logging
No external dependencies required
"""

import os
import re
import sys
from pathlib import Path

def find_typescript_files(root_path):
    """Find all TypeScript files excluding tests"""
    files = []
    for root, dirs, filenames in os.walk(root_path):
        # Skip node_modules
        if 'node_modules' in root:
            continue
        for filename in filenames:
            if filename.endswith('.ts') and not filename.endswith('.test.ts') and not filename.endswith('.spec.ts'):
                files.append(os.path.join(root, filename))
    return files

def replace_console_statements(content):
    """Replace console.* with logStructuredEvent"""
    changes = 0
    
    # Pattern for console.log
    content, count = re.subn(
        r'console\.log\((.*?)\);?',
        r'logStructuredEvent("DEBUG", { data: \1 });',
        content
    )
    changes += count
    
    # Pattern for console.error
    content, count = re.subn(
        r'console\.error\((.*?)\);?',
        r'logStructuredEvent("ERROR", { error: \1 }, "error");',
        content
    )
    changes += count
    
    # Pattern for console.warn
    content, count = re.subn(
        r'console\.warn\((.*?)\);?',
        r'logStructuredEvent("WARNING", { message: \1 }, "warn");',
        content
    )
    changes += count
    
    return content, changes

def add_import_if_needed(content):
    """Add logStructuredEvent import if not present"""
    if 'logStructuredEvent' in content:
        return content
    
    # Find the last import statement
    import_pattern = r'(import.*from.*[\'"];?\n)'
    matches = list(re.finditer(import_pattern, content))
    
    if matches:
        last_import = matches[-1]
        insert_pos = last_import.end()
        import_statement = 'import { logStructuredEvent } from "../_shared/observability.ts";\n'
        content = content[:insert_pos] + import_statement + content[insert_pos:]
    else:
        # No imports, add at top
        import_statement = 'import { logStructuredEvent } from "../_shared/observability.ts";\n\n'
        content = import_statement + content
    
    return content

def process_file(filepath, dry_run=False):
    """Process a single file"""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            original_content = f.read()
        
        # Replace console statements
        new_content, changes = replace_console_statements(original_content)
        
        if changes > 0:
            # Add import if needed
            new_content = add_import_if_needed(new_content)
            
            if not dry_run:
                with open(filepath, 'w', encoding='utf-8') as f:
                    f.write(new_content)
            
            return filepath, changes
        
        return None, 0
    except Exception as e:
        print(f"  ❌ Error processing {filepath}: {e}", file=sys.stderr)
        return None, 0

def main():
    dry_run = '--dry-run' in sys.argv
    path_arg = next((arg for arg in sys.argv if arg.startswith('--path=')), None)
    search_path = path_arg.split('=')[1] if path_arg else 'supabase/functions/wa-webhook-mobility'
    
    print(f"🔍 Scanning {search_path} for console.log statements...")
    print(f"   Mode: {'DRY RUN' if dry_run else 'LIVE FIX'}\n")
    
    # Find all TypeScript files
    files = find_typescript_files(search_path)
    print(f"   Found {len(files)} TypeScript files\n")
    
    # Process files
    total_replacements = 0
    modified_files = []
    
    for filepath in files:
        result, changes = process_file(filepath, dry_run)
        if result:
            total_replacements += changes
            modified_files.append((result, changes))
    
    # Report results
    print(f"\n{'=' * 70}")
    print(f"📊 SUMMARY")
    print(f"{'=' * 70}")
    print(f"Files scanned:      {len(files)}")
    print(f"Files modified:     {len(modified_files)}")
    print(f"Total replacements: {total_replacements}")
    print(f"{'=' * 70}\n")
    
    if modified_files:
        print(f"📝 Changes:\n")
        for filepath, changes in modified_files:
            rel_path = os.path.relpath(filepath)
            print(f"  ✓ {rel_path} ({changes} replacement{'s' if changes > 1 else ''})")
        print()
    
    if dry_run:
        print(f"⚠️  DRY RUN MODE - No files were modified")
        print(f"   Run without --dry-run to apply changes\n")
    else:
        print(f"✅ Changes applied successfully!\n")
    
    return 0 if not dry_run or total_replacements == 0 else 1

if __name__ == '__main__':
    sys.exit(main())

File: scripts/fix-console-logs.mjs
#!/usr/bin/env node

/**
 * Replace console.log with structured logging
 * 
 * Usage:
 *   node scripts/fix-console-logs.mjs [--dry-run] [--path=supabase/functions/wa-webhook-mobility]
 */

import { readFileSync, writeFileSync } from 'fs';
import { glob } from 'glob';
import { dirname, relative } from 'path';

const args = process.argv.slice(2);
const dryRun = args.includes('--dry-run');
const pathArg = args.find(a => a.startsWith('--path='));
const searchPath = pathArg ? pathArg.split('=')[1] : 'supabase/functions/wa-webhook-mobility';

console.log(`🔍 Scanning ${searchPath} for console.log statements...`);
console.log(`   Mode: ${dryRun ? 'DRY RUN' : 'LIVE FIX'}\n`);

// Find all TypeScript files (excluding tests)
const files = glob.sync(`${searchPath}/**/*.ts`, {
  ignore: ['**/*.test.ts', '**/*.spec.ts', '**/node_modules/**']
});

let totalFiles = 0;
let totalReplacements = 0;
const changes = [];

files.forEach(file => {
  const content = readFileSync(file, 'utf-8');
  let modified = content;
  let fileReplacements = 0;

  // Pattern 1: console.log(...)
  const consoleLogRegex = /console\.log\((.*?)\);?/g;
  
  // Pattern 2: console.error(...)
  const consoleErrorRegex = /console\.error\((.*?)\);?/g;
  
  // Pattern 3: console.warn(...)
  const consoleWarnRegex = /console\.warn\((.*?)\);?/g;

  // Check if file already imports logStructuredEvent
  const hasImport = /import.*logStructuredEvent.*from/.test(content);
  
  // Replace console.log
  if (consoleLogRegex.test(content)) {
    modified = modified.replace(consoleLogRegex, (match, args) => {
      fileReplacements++;
      
      // Try to extract meaningful event name from the log
      const firstArg = args.split(',')[0].trim();
      let eventName = 'LOG_EVENT';
      
      if (firstArg.includes('DEBUG')) {
        eventName = 'DEBUG';
      } else if (firstArg.includes('loaded') || firstArg.includes('Loaded')) {
        eventName = 'MODULE_LOADED';
      } else if (firstArg.includes('called') || firstArg.includes('Called')) {
        eventName = 'FUNCTION_CALLED';
      }
      
      // If single string argument, make it the event name
      if (!args.includes(',') && firstArg.startsWith('"') && firstArg.endsWith('"')) {
        return `logStructuredEvent(${firstArg}, {});`;
      }
      
      // Multiple arguments - first is event, rest is payload
      const [first, ...rest] = args.split(',').map(s => s.trim());
      if (rest.length > 0) {
        return `logStructuredEvent("${eventName}", { message: ${first}, data: ${rest.join(', ')} });`;
      }
      
      return `logStructuredEvent("${eventName}", { data: ${args} });`;
    });
  }

  // Replace console.error
  if (consoleErrorRegex.test(content)) {
    modified = modified.replace(consoleErrorRegex, (match, args) => {
      fileReplacements++;
      return `logStructuredEvent("ERROR", { error: ${args} }, "error");`;
    });
  }

  // Replace console.warn
  if (consoleWarnRegex.test(content)) {
    modified = modified.replace(consoleWarnRegex, (match, args) => {
      fileReplacements++;
      return `logStructuredEvent("WARNING", { message: ${args} }, "warn");`;
    });
  }

  // Add import if needed and changes were made
  if (fileReplacements > 0 && !hasImport) {
    // Find the right path to observability.ts based on file location
    const relPath = relative(dirname(file), searchPath);
    const depth = relPath.split('/').filter(p => p === '..').length || 1;
    const importPath = '../'.repeat(depth) + '_shared/observability.ts';
    
    // Add import after other imports
    const importStatement = `import { logStructuredEvent } from "${importPath}";\n`;
    
    // Find last import statement
    const lastImportMatch = modified.match(/import.*from.*;\n/g);
    if (lastImportMatch) {
      const lastImport = lastImportMatch[lastImportMatch.length - 1];
      modified = modified.replace(lastImport, lastImport + importStatement);
    } else {
      // No imports, add at top
      modified = importStatement + modified;
    }
  }

  if (fileReplacements > 0) {
    totalFiles++;
    totalReplacements += fileReplacements;
    changes.push({
      file: relative(process.cwd(), file),
      replacements: fileReplacements
    });

    if (!dryRun) {
      writeFileSync(file, modified, 'utf-8');
    }
  }
});

// Report results
console.log(`\n${'='.repeat(70)}`);
console.log(`📊 SUMMARY`);
console.log(`${'='.repeat(70)}`);
console.log(`Files scanned:     ${files.length}`);
console.log(`Files modified:    ${totalFiles}`);
console.log(`Total replacements: ${totalReplacements}`);
console.log(`${'='.repeat(70)}\n`);

if (changes.length > 0) {
  console.log(`📝 Changes:\n`);
  changes.forEach(({ file, replacements }) => {
    console.log(`  ✓ ${file} (${replacements} replacement${replacements > 1 ? 's' : ''})`);
  });
  console.log();
}

if (dryRun) {
  console.log(`⚠️  DRY RUN MODE - No files were modified`);
  console.log(`   Run without --dry-run to apply changes\n`);
} else {
  console.log(`✅ Changes applied successfully!\n`);
}

process.exit(0);
